From ee2ece01a5a36cbfc5c25d6c3f9ded72f0e912c0 Mon Sep 17 00:00:00 2001
From: Gustavo Duarte <gduarte@redhat.com>
Date: Mon, 13 Oct 2014 22:30:05 -0400
Subject: [misc] genwqe: Add wmb before DDCB is started

Message-id: <1413239416-7778-18-git-send-email-gduarte@redhat.com>
Patchwork-id: 97558
O-Subject: [RHEL7.1 PATCH BZ 1088491 17/28 v2] GenWQE: Add wmb before DDCB is started
Bugzilla: 1088491
RH-Acked-by: Don Zickus <dzickus@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1088491
Upstream Status: 68fe8acc204c7fbefd4c01b8929fedb244ec283d

commit 68fe8acc204c7fbefd4c01b8929fedb244ec283d
Author: Frank Haverkamp <haver@linux.vnet.ibm.com>
Date:   Thu Mar 20 15:11:03 2014 +0100

    GenWQE: Add wmb before DDCB is started

    Needed to add wmb() before we send the DDCB for execution.
    Without the syncronizing it failed on System p.

    Signed-off-by: Frank Haverkamp <haver@linux.vnet.ibm.com>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/misc/genwqe/card_ddcb.c b/drivers/misc/genwqe/card_ddcb.c
index 6f1acc0..29a1a28 100644
--- a/drivers/misc/genwqe/card_ddcb.c
+++ b/drivers/misc/genwqe/card_ddcb.c
@@ -305,6 +305,8 @@ static int enqueue_ddcb(struct genwqe_dev *cd, struct ddcb_queue *queue,
    break;
 
   new = (old | DDCB_NEXT_BE32);
+
+  wmb();
   icrc_hsi_shi = cmpxchg(&prev_ddcb->icrc_hsi_shi_32, old, new);
 
   if (icrc_hsi_shi == old)
@@ -314,6 +316,8 @@ static int enqueue_ddcb(struct genwqe_dev *cd, struct ddcb_queue *queue,
  /* Queue must be re-started by updating QUEUE_OFFSET */
  ddcb_mark_tapped(pddcb);
  num = (u64)ddcb_no << 8;
+
+ wmb();
  __genwqe_writeq(cd, queue->IO_QUEUE_OFFSET, num); /* start queue */
 
  return RET_DDCB_TAPPED;
-- 
1.7.1